### PR TITLE
Fix for GEODE-3898

### PR DESCRIPTION
**Pull request by jujoramos**
GEODE-3898: Defined Indexes are not Persisted in Cluster Configuration.
  . Added JUnit tests.
  . CreateDefinedIndexesFunction returns a list of CliFunctionResult and XmlEntity with the indexes created.
  . CreateDefinedIndexesCommand parses the list of results and persists the created indexes in the cluster configuration service.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [X] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/CreateDefinedIndexesCommand.java [hunk 1]
```diff
@@ -15,13 +15,13 @@
 
 package org.apache.geode.management.internal.cli.commands;
 
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
-import java.util.TreeSet;
+import java.util.*;
```
Comment (jdeppe-pivotal) [inline geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/CreateDefinedIndexesCommand.java]: Please use explicit imports. Preferably have your IDE use the relevant style guide in `etc/`.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/CreateDefinedIndexesCommand.java [hunk 2]
```diff
@@ -89,23 +88,25 @@ public Result createDefinedIndexes(
...
             failedMembers.add(cliFunctionResult.getMemberIdOrName());
             indexOpFailMap.put(exceptionMessage, failedMembers);
           }
         }
       }
 
+      // TODO:
+      // with the current logic some indexes might be correctly created in some members and
+      // fail in others; but the user is never notified about these failures if there's at least one
+      // successful creation...
```
Comment (jdeppe-pivotal) [inline geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/CreateDefinedIndexesCommand.java]: I'm not sure how consistent we are on this, but I think the command should fail if there are **any** errors at all. I know that we do this in `CreateRegionCommand`. We're adding options to allow for idempotency in all commands, so cleaning up from a partial failure will become very easy. i.e. on failure you can then issue: `destroy index --if-exists`

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/CreateDefinedIndexesFunction.java [hunk 3]
```diff
@@ -14,39 +14,57 @@
...
 import org.apache.geode.cache.query.IndexType;
 import org.apache.geode.cache.query.MultiIndexCreationException;
 import org.apache.geode.cache.query.QueryService;
 import org.apache.geode.internal.InternalEntity;
+import org.apache.geode.internal.cache.xmlcache.CacheXml;
+import org.apache.geode.internal.logging.LogService;
 import org.apache.geode.management.internal.cli.domain.IndexInfo;
 import org.apache.geode.management.internal.cli.i18n.CliStrings;
+import org.apache.geode.management.internal.configuration.domain.XmlEntity;
+import org.apache.logging.log4j.Logger;
```
Comment (jdeppe-pivotal) [inline geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/CreateDefinedIndexesFunction.java]: Some imports are unused.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/CreateDefinedIndexesFunction.java [hunk 4]
```diff
@@ -14,39 +14,57 @@
...
-
   private static final long serialVersionUID = 1L;
 
+
+  @Override
+  public String getId() {
+    return CreateDefinedIndexesFunction.class.getName();
+  }
+
+  // For mocking purposes.
```
Comment (jdeppe-pivotal) [inline geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/CreateDefinedIndexesFunction.java]: Apparently not just for mocking :)

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/CreateDefinedIndexesCommand.java (hunk 1 above)
Comment (jujoramos) [reply geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/CreateDefinedIndexesCommand.java]: Thanks for the tip about the style, I'll include the change in the next commit.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/CreateDefinedIndexesCommand.java (hunk 2 above)
Comment (jujoramos) [reply geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/CreateDefinedIndexesCommand.java]: I'm not 100% sure about this one... currently creation can be successful on some members and fail on others, in that case we should report to the user which indexes were successfully created and which ones failed, adding the details about the members on which the operation failed/succeeded. I've created GEODE-3916 to address this.
Regarding this pull request, I'll add more details in the comment for future reference.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/CreateDefinedIndexesFunction.java (hunk 3 above)
Comment (jujoramos) [reply geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/CreateDefinedIndexesFunction.java]: Thanks for the tip.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/CreateDefinedIndexesFunction.java (hunk 4 above)
Comment (jujoramos) [reply geode-core/src/main/java/org/apache/geode/management/internal/cli/functions/CreateDefinedIndexesFunction.java]: True, I just tried to make it clear that I externalized the method to be able to mock the behavior from the outside. Comment removed :-).

**Comment by jujoramos** [comment]
Hey @jinmeiliao, thanks for having a look at this.
I might be missing something, but within `CreateIndexCommand` you can just persist one `XmlEntity` to the cluster configuration service because you only have one index created on a single region.
In this particular case, `CreateDefinedIndexesCommand` can create more than one index and not necessarily for the same region, that's why all of the  `XmlEntity` should be persisted to the cluster configuration service.
As an example:

```
gfsh -e "connect --locator=localhost[10101]" \
	-e "create region --name=TestRegion1 --type=REPLICATE" \
	-e "create region --name=TestRegion2 --type=REPLICATE"
gfsh -e "connect --locator=localhost[10101]" \
	-e "define index --name=index1 --expression=value1 --region=TestRegion1" \
	-e "define index --name=index2 --expression=value2 --region=TestRegion2"

gfsh -e "connect --locator=localhost[10101]" -e "create defined indexes"
```

In the above case, the server will need to return two `CliFunctionResult` with two `XmlEntity` instances (one for each region), and the `CreateDefinedIndexesCommand` must persist both entities to the cluster configuration service to make sure we don't loose anything:

```
XmlEntity [namespace=http://geode.apache.org/schema/cache, type=region, attributes={name=TestRegion2}, xmlDefinition=<?xml version="1.0" encoding="UTF-8"?><region name="TestRegion2" xmlns="http://geode.apache.org/schema/cache">
      <region-attributes data-policy="replicate" scope="distributed-ack"/>
      <index expression="value2" from-clause="/TestRegion2" key-index="false" name="index2" type="range"/>
    </region>]

XmlEntity [namespace=http://geode.apache.org/schema/cache, type=region, attributes={name=TestRegion1}, xmlDefinition=<?xml version="1.0" encoding="UTF-8"?><region name="TestRegion1" xmlns="http://geode.apache.org/schema/cache">
      <region-attributes data-policy="replicate" scope="distributed-ack"/>
      <index expression="value1" from-clause="/TestRegion1" key-index="false" name="index1" type="range"/>
    </region>]
```

Am I missing something?.

**Comment by jinmeiliao** [comment]
@jujoramos oh yes, you are right. It could possibly creating multiple indexes. Is it ever possible that different member might return the same xml for the same index if both host the region? Have you tested this scenario? It might be good to add a DUnit test for this. Consider using LocatorServerStarterRule to start a locator and multiple servers that has the same region, see if the CC is updated correctly with just one entry.

**Comment by jujoramos** [comment]
Hey @jinmeiliao ,

I've tested the scenario but never added the DUnit test :-/.
According to my understanding the regions must have the same configuration on every member where it's hosted, so in the worst case scenario we'll end up persisting the same configuration for the region `N` times, with `N = #membersHostingTheRegion`.
The change will imply changing `CreateDefinedIndexesCommand` to persist the cluster configuration **only once per region whenever at least one member was able to succesfully create the index on the region**, something like:

```
...

for (final CliFunctionResult funcResult : funcResults) {
	if (cliFunctionResult.isSuccessful()) {
		...

		// Only add the XmlEntity if it wasn't previously added from the result of another successfull member. XmlEntity.equals() is already implemented.
		if (!xmlEntities.contains(xmlEntities)) {
			xmlEntities.add(cliFunctionResult.getXmlEntity());
		}
		...
	}
}

...

if (!xmlEntities.isEmpty()) {
	for (XmlEntity xmlEntity : xmlEntities) {
		persistClusterConfiguration(result,
			() -> getSharedConfiguration().addXmlEntity(xmlEntity.get(), group));
	}
}

...
```

Should I make the changes and add the new DUnit test within the current pull request or directly open a new JIRA and add the changes in a new pull request relevant to that JIRA?.

**Comment by jinmeiliao** [comment]
I would suggest making the change now so that we wouldn't persisting the same index multiples times in CC and add the tests to verify that. Thanks!

**Comment by jujoramos** [comment]
Hey @jinmeiliao ,

I've made the requested changes, compiled and tested everything locally, and pushed.
Now the task `:geode-core:compileTestJava` is failing within Travis, with errors similar the one below:

```
error: cannot find symbol gfsh.executeAndVerifyCommand("", "");
        ^
  symbol:   method executeAndVerifyCommand(String,String)
```

The import statements are correct, the method from the rule is being called from several other DUnit tests, and everything works fine locally... any clues why it's failing on Travis?.

**Comment by jinmeiliao** [comment]
You will need to rebase to the current develop. That method is being refactored out.

**Comment by jujoramos** [comment]
I'm closing this pull request, will create a new one containing a single commit with all the changes.

**Pull request closed without merging**